annot/src/main/java/com/predic8/membrane/annot/generator/JsonSchemaGenerator.java

```diff
@@ -13,46 +13,87 @@
-import javax.tools.FileObject;
-import javax.tools.StandardLocation;
-import java.io.BufferedWriter;
-import java.io.IOException;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.stream.Collectors;
-
+import com.predic8.membrane.annot.model.doc.*;
+
+import javax.annotation.processing.*;
+import javax.lang.model.element.*;
+import javax.tools.*;
+import java.io.*;
+import java.util.*;
+
+import static com.predic8.membrane.annot.generator.kubernetes.model.SchemaFactory.*;
+import static com.predic8.membrane.annot.generator.kubernetes.model.SchemaFactory.ref;
+import static com.predic8.membrane.annot.generator.util.SchemaGeneratorUtil.*;
+import static javax.tools.StandardLocation.*;
+
+/**
+ * TODOs:
+ * - A required property with a base type needs all the subtypes to be present in the schema. See CacheParser
+ * - ports are strings
+ * - Choose/cases/case has one nesting to much
+ * - apiKey/extractors/expressionExtractor/expression => too much?
+ */
 public class JsonSchemaGenerator extends AbstractK8sGenerator {
 
+    private final Map<String,Boolean> topLevelAdded = new HashMap<>();
+
     public JsonSchemaGenerator(ProcessingEnvironment processingEnv) {
         super(processingEnv);
     }
 
-    public void write(Model m) {
-        try {
-            for (MainInfo main : m.getMains()) {
-                assemble(m, main);
-            }
-        } catch (IOException e) {
-            e.printStackTrace();
+    private boolean flowDefCreated = false;
+    private Schema schema;
+
+    private static final Set<String> excludeFromFlow = Set.of(
+            "httpClient",
+            "ruleMatching",
+            "wadlRewriter",
+            "global",
+            "exchangeStore",
+            "accountRegistration",
+            "userFeature",
+            "tcp",
+            "wsaEndpointRewriter",
+            "flowInitiator",
+            "kubernetesValidation",
+            "dispatching",
+            "groovyTemplate",
+            "adminApi"
+    );
+
+    public void write(Model m) throws IOException {
+        for (MainInfo main : m.getMains()) {
+            assemble(m, main);
         }
     }
 
     private void assemble(Model m, MainInfo main) throws IOException {
-        Schema schema = new Schema("membrane");
-        List<RefObj> oneOfArray = new ArrayList<>();
+        // Reset so multiple calls would be possible
+        flowDefCreated = false;
+        schema = schema("membrane");
+        topLevelAdded.clear();
 
+        addParserDefinitions(m, main);
+        addTopLevelProperties();
+
+        writeSchema(main, schema);
+    }
+
+    private void addTopLevelProperties() {
+        schema.additionalProperties(false)
+            .property(string("apiVersion"))
+            .property( string("kind").enumeration(List.of("api")))
+            .property( ref("spec").ref("#/$defs/com.predic8.membrane.core.config.spring.ApiParser").required(true))
+            .property(object("metadata")
```

Comment:
on 'metadata', we should set `.additionalProperties(true)`: Kubernetes has a variety of additional properties below metadata, which are of no interest to us.